src/Language/PureScript/CoreFn/Optimizer.hs

```diff
@@ -48,9 +51,17 @@ closedRecordFields (TypeApp _ (TypeConstructor _ C.Record) row) =
 closedRecordFields _ = Nothing
 
 -- | See https://github.com/purescript/purescript/issues/3157
-optimizeUnusedPartialFn :: Expr a -> Expr a
+optimizeUnusedPartialFn :: Expr Ann -> Expr Ann
 optimizeUnusedPartialFn (Let _
   [NonRec _ UnusedIdent _]
   (App _ (App _ (Var _ (Qualified _ UnusedIdent)) _) originalCoreFn)) =
   originalCoreFn
 optimizeUnusedPartialFn e = e
+
+-- | TODO: Fixup the annotations here.
+optimizeDataFunctionApply :: Expr Ann -> Expr Ann
+optimizeDataFunctionApply e = case e of
+  (App a (App _ (Var _ (Qualified (Just (ModuleName dataFunction)) (Ident applyFn))) x) y)
+    | dataFunction == "Data.Function" && applyFn == "apply" -> App a x y
+    | dataFunction == "Data.Function" && applyFn == "applyFlipped" -> App a y x
```

Comment:
It seems like this is enough to inline most usages of `apply`/`applyFlipped`.

Reply:
Replace the string literals with constants from `Language.PureScript.Constants.Prelude`, please. (This will also resolve your Weeder errors; the unused declarations it's identifying are the ones you need to use.)